src/pgas.jl

```diff
@@ -0,0 +1,201 @@
+end
+
+"""
+    get_ancestor_logweights(pc::ParticleContainer{F,R}, x) where {F<:SSMTrace,R}
+
+Get the ancestor log weights for each particle in `pc`
+"""
+function get_ancestor_logweights(pc::ParticleContainer{F,R}, x) where {F<:SSMTrace,R}
+    particles = collect(pc.vals)
+    nparticles = length(particles)
+    weights = pc.logWs
+
+    logweights = map(1:nparticles) do i
+        transition_logweight(particles[i], x) + weights[i]
+    end
+
+    return logweights
+end
+
+""" 
+    advance!(particle::SSMTrace, isref::Bool=false)
+
+Return the log-probability of the transition nothing if done
+"""
+function advance!(particle::SSMTrace, isref::Bool=false)
+    isref ? load_state!(particle.rng) : save_state!(particle.rng)
+
+    model = particle.f
+    if !isref
+        if model.T == 1
+            new_state = rand(particle.rng, initialization(model)) # Generate initial state, maybe fallback to 0 is initialization is not defined
+        else
+            current_state = model.X[model.T - 1]
+            new_state = rand(particle.rng, transition(model, current_state))
+        end
+    else
+        new_state = model.X[model.T <= length(model.X) ? model.T : end] # Hacky, reached beyond last time step
+    end
+
+    score = observation(model, new_state)
+
+    if score !== nothing # Accept transition
+        !isref && push!(model.X, new_state)
+        inc_step!(model)
+        inc_counter!(particle.rng) # Increase rng counter, don't really need it but keep things consistent
+    else
+        inc_step!(model, -1)
+    end
+
+    return score
+end
+
+function truncate!(model::AbstractStateSpaceModel)
+    step = model.T <= length(model.X) ? model.T - 1 : length(model.X)
+    return model.X = model.X[1:step]
+end
+
+function fork(trace::SSMTrace, isref::Bool)
+    model = deepcopy(trace.f)
+    isref && truncate!(model) # Forget the rest of the reference trajectory
+    return SSMTrace(model, trace.rng)
+end
+
+function forkr(trace::SSMTrace)
+    Random123.set_counter!(trace.rng, 1)
+    reset!(trace.f) # Reset model count, we could probably use the TracedRNG inner counter 
+    return SSMTrace(deepcopy(trace.f), trace.rng)
+end
+
+function update_ancestor!(ref::SSMTrace, pc::ParticleContainer)
+    current_step(ref.f) == 1 && return nothing
+
+    particles = collect(pc.vals)
+    ancestor_weights = get_ancestor_logweights(pc, ref.f.X[ref.f.T - 1])
+    weights = StatsFuns.softmax(ancestor_weights)
+    ancestor_index = rand(Distributions.Categorical(weights))
+    ancestor = particles[ancestor_index]
+    return ref.f.X[ref.f.T - 1] = previous_state(ancestor.f)
+end
+
+function resample_propagate!(
+    rng::Random.AbstractRNG,
+    pc::ParticleContainer{T,R},
+    randcat=resample_systematic,
+    ref::Union{Particle,Nothing}=nothing;
+    weights=getweights(pc),
+) where {T<:SSMTrace,R}
+
+    # check that weights are not NaN
+    @assert !any(isnan, weights)
```

Comment:
Is this needed? It seems expensive (have to go through all weights in every resampling step) and I assume we would just end up with `NaN` or some other error anyway.

Reply:
Not really, computing the weights would throw an error. That's also in the old resampling step:
https://github.com/TuringLang/AdvancedPS.jl/blob/9f8522f4e2219430255c89f41aa990c2b93104c2/src/container.jl#L230-L231
I can remove these